### ISSUE TITLE
[NET8 Log Reduction] Detect direct logging of reducted information

### ISSUE THREAD

**Issue by LiorBanai** [meziantou/Meziantou.Analyzer#670]
**Is your feature request related to a problem? Please describe.**

in NET8 Microsoft added recution capabilites for logging  exampel: https://andrewlock.net/redacting-sensitive-data-with-microsoft-extensions-compliance/

**Describe the solution you'd like**

I would like the have analyzer that detects direct logging of proeprties that uses reduction attributes.

**Describe alternatives you've considered**

There could be a step during compilation but if it i spossible to detect it as soon as possible it wloud be great.

**Additional context**

N/A

Greate library. I"m using it in all off my projects

**Comment by meziantou** [comment]
Can you provide a small sample of what should be detected?

**Comment by LiorBanai** [comment]
let say I do not want to log patient infomation in logs (but still have log patient data) like:
```csharp
   public partial class PatientInfo
   {
        [PiiData] public string PatientId { get; set; }
        public ulong RecordId { get; set; } //the record Id in the backend database
        [PiiData] public string FirstName { get; set; }
        [PiiData] public string MiddleName { get; set; }
        [PiiData] public string LastName { get; set; }
        public DateTimeOffset? ScheduledDate { get; set; } //UTC value
        public string Department { get; set; }
        public DateTimeOffset? DateOfBirth { get; set; } //UTC value
        [PiiData] public string MedicalRecordNumber { get; set; } //Same as Patient ID
}
```

when using reduction and using the Log Properties attribute as:
```csharp
        [LoggerMessage(Level = LogLevel.Information, Message = "Patient Information {source}")]
        public static partial void LogPatient(ILogger logger, string source, [LogProperties] PatientInfo patient);
```

the properties with ```[PiiData]``` are reducted using:

```csharp
          services.AddRedaction(x =>
          {
              x.SetRedactor<StarRedactor>(new DataClassificationSet(DataTaxonomy.SensitiveData));
              x.SetRedactor<StarRedactor>(new DataClassificationSet(DataTaxonomy.PiiData));
          });
```

which in the log is shown as:
```json
{
	"@timestamp": "2023-12-22T18:18:26.0423257+02:00",
	"log.level": "Information",
	"message": "Patient Information method1",
	"ecs.version": "8.6.0",

	"labels": {
		"MessageTemplate": "Patient Information {source}",
		"patient.DateOfBirth": "24/06/1997 7:57:48 +00:00",
		"patient.Department": "Surgery",
		"patient.ScheduledDate": "21/12/2023 6:49:44 +02:00",
		"patient.MedicalRecordNumber": "**************************************",
		"patient.LastName": "**********************",
		"patient.MiddleName": "************************",
		"patient.FirstName": "**************************",
		"patient.PatientId": "***********************"
	}
}
```

but if I log direct properties like
```csharp
Logger.LogInformation("Patient name: {name}", patient.FirstName);
```

that will be logged. I would like to get a warning when I try to log any PII information ( or any other reducted properties) like in the above line.

**Comment by meziantou** [comment]
If I well understood, the new rule should detect the usage of any member (field, property, parameter) decorated with `Microsoft.Extensions.Compliance.Classification.DataClassificationAttribute`

````c#
// Report the second argument
logger.LogInformation("dummy", obj.PropertyDecoratedWithDataClassificationAttribute);
````

**Comment by LiorBanai** [comment]
Thanks for the fast solution. I'll give it a try and update

**Comment by LiorBanai** [comment]
@meziantou I Just wanted to inform you that this rule just detected incorrect logging of sensitive data by a team member (we set it to an error)
![image](https://github.com/user-attachments/assets/7c753bac-0d85-4d77-957a-bb6eff7dce0b)

thanks a lot for your great work

**Comment by meziantou** [comment]
@LiorBanai Can you provide a repro?

**Comment by LiorBanai** [comment]
@meziantou I meant a team member incorrectly logged sensitive data and your analyzer detect it correctly and error out.

I worded it badly. Your analyzer works excatly as desired (and warns us of forbidden logging). 

Thanks  for all your work.

**Comment by LiorBanai** [comment]
@meziantou is it possible to update the rule to check top level attribute? aka:
```csharp
[PiiData]  
 public partial class PatientInfo
   {
        public string PatientId { get; set; }
        public ulong RecordId { get; set; } //the record Id in the backend database
        public string FirstName { get; set; }
        public string MiddleName { get; set; }
        public string LastName { get; set; }
        public DateTimeOffset? ScheduledDate { get; set; } //UTC value
        public string Department { get; set; }
        public DateTimeOffset? DateOfBirth { get; set; } //UTC value
        public string MedicalRecordNumber { get; set; } //Same as Patient ID
}
PatientInfo patient = new..
Logger.LogInformation("Patient data: {name}", patient);
```

right now this does not produce error/warning

**Comment by meziantou** [comment]
@LiorBanai It will be available in the next version.

Next time, you should open a new issue instead of adding a comment on a closed one.
